### PR TITLE
8278508: Enable X86 maskAll instruction pattern for 32 bit JVM.

### PR DESCRIPTION
**Pull request by jatin-bhateja**
- Vector.maskAll was accelerated for AVX-512 target, but  x86 existing backend implementation does not enable maskAll instruction patterns for 32 bit JVM, due to which operations fall backs over replicateB operation which broadcasts the mask value in a vector.
- In some cases after unboxing-boxing optimization this vector eventually reaches to XorVMask which has different operands one held in opmask register and other in vector.

Kindly review and share your feedback.

Best Regards,
Jatin

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8278508](https://bugs.openjdk.java.net/browse/JDK-8278508): Enable X86 maskAll instruction pattern for 32 bit JVM.

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 4fb1ea1b1209d383384953dfcab067064fc60e7f
 * [Sandhya Viswanathan](https://openjdk.java.net/census#sviswanathan) (@sviswa7 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk18 pull/24/head:pull/24` \
`$ git checkout pull/24`

Update a local copy of the PR: \
`$ git checkout pull/24` \
`$ git pull https://git.openjdk.java.net/jdk18 pull/24/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 24`

View PR using the GUI difftool: \
`$ git pr show -t 24`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk18/pull/24.diff">https://git.openjdk.java.net/jdk18/pull/24.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
Seems reasonable. Let me test it.

**Review by vnkozlov — CHANGES_REQUESTED** [review]
Regular testing (x64) results are good.

But I also run `jdk/incubator/vector/` tests locally with 32-bit VM (fastdebug). Most tests passed but some hit timeout (run for more than default 2 min):
```
jdk/incubator/vector/Byte512VectorTests.java
jdk/incubator/vector/ByteMaxVectorTests.java
jdk/incubator/vector/VectorReshapeTests.java
```
I assume such vectors are not supported in 32-bit VM and code is slow. Simple solution is to add `/timeout=240` (4 min) to these tests. They passed for me after that.

**Comment by jatin-bhateja** [comment]
> Regular testing (x64) results are good.
> 
> But I also run `jdk/incubator/vector/` tests locally with 32-bit VM (fastdebug). Most tests passed but some hit timeout (run for more than default 2 min):
> 
> ```
> jdk/incubator/vector/Byte512VectorTests.java
> jdk/incubator/vector/ByteMaxVectorTests.java
> jdk/incubator/vector/VectorReshapeTests.java
> ```
> 
> I assume such vectors are not supported in 32-bit VM and code is slow. Simple solution is to add `/timeout=240` (4 min) to these tests. They passed for me after that.

Hi @vnkozlov , x86 32bit target constraints the number of vector registers to 8 (zmm0-zmm7) so above testcases should still be able to intrinsify the operation. I tested above tests on AVX512 target and they executes in lass than 2 mins. On a non-AVX512 target any operation based on Byte512 species is not inline expanded. 

Since the changes in this patch are only related to AVX512 thus my testing went clean with UseAVX=3.

**Comment by vnkozlov** [comment]
I ran these tests with `-XX:-TieredCompilation -Xbatch -XX:UseAVX=3` flags and 32-bit VM on Skylake.

**Comment by jatin-bhateja** [comment]
> ```
> jdk/incubator/vector/VectorReshapeTests.java
> ```

Hi @vnkozlov , I have added explicit time out check as suggested in above 3 testcases.

**Review by vnkozlov — APPROVED** [review]
Good. You need second review for this.

**Comment by jatin-bhateja** [comment]
@svisva7 can you kindly be second reviewer.

File: src/hotspot/cpu/x86/x86_32.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86_32.ad]: This can be removed. The more generic  mask_all_evexI_GT32 rule will take care of this.

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86.ad]: TEMP dst is not needed.

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86.ad]: TEMP dst is not needed here.

File: src/hotspot/cpu/x86/x86_32.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86_32.ad]: TEMP dst is not needed.

File: src/hotspot/cpu/x86/x86_64.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86_64.ad]: TEMP dst not needed.

File: src/hotspot/cpu/x86/x86_64.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86_64.ad]: TEMP dst not needed.

File: src/hotspot/cpu/x86/x86_64.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86_64.ad]: TEMP dst not needed.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: If masklen == 16 then kshiftrwl is not needed?

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: Here masklen is 64, so kshiftrql is not needed here?

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: If masklen is 32 then kshiftrdl is not needed. Only needed if masklen < 32.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: Do we need the kshiftrql here? The masklen is 64 here.
You could alternatively use:
   kmovdl dst, src
   kunpckdq dst, dst, dst

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: src operand carry either a 0 (false mask) or 0XFFFFFFFFFFFFFFFF  (true mask) value. Thus a right shift here ensures that destination bits whose size matches with mask_len are set.

I have introduced special case handling when mask_len is  64, 32 and 16.

File: src/hotspot/cpu/x86/x86.ad
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/x86.ad]: dst is both read and written to in macro assembly routine.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: DONE

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: We are only supporting vector lengths of power of 2 for x86 (8,16,32,64) for byte vector.  The only case that comes here is mask_len == 64.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: We are only supporting vector lengths of power of 2 for x86 (8,16,32,64) for byte vector. The only case that comes here is mask_len == 64 so we dont need kshiftrql.

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [reply src/hotspot/cpu/x86/x86.ad]: dst both written and read is not the criteria for TEMP dst effect. Only if the src is used after dest is written in the encoding for instruct, TEMP dst is needed.

Comment (merykitty) [reply src/hotspot/cpu/x86/x86.ad]: Hi, do we need `TEMP dst` if some `tmp` is used after `dst` is written, as without it `tmp` and `dst` are implied to have non-overlapping lifetime and can lead to conflict. This is a pure question unrelated to the PR. Thank you very much.

Comment (sviswa7) [reply src/hotspot/cpu/x86/x86.ad]: Hi @merykitty, yes we would need TEMP dst in that case as well.

Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/x86.ad]: TEMP attribute ensures creation of a temporary machine operand which interferes with source operands if their registerclasses are overlapping, a TEMP + DST data flow attribute ensures that DST is not accidentally allocated a SRC register even if SRC is not live beyond that instruction. This shall prevent overriding the src before using its value in the instruction encoding block.

**Comment by jatin-bhateja** [comment]
Hi @sviswa7, your comments addressed. TEMP def was  an artifact of some other change I was doing initially.

**Review by sviswa7 — APPROVED** [review]
Patch looks good to me.

**Pull request closed without merging**
